### ISSUE TITLE
[java] New rule: AvoidPublicNativeMethods

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2986]
**Proposed Rule Name:**
AvoidPublicNativeMethods

**Proposed Category:**  
Security

**Description:**
Native methods must not be declared public but wrapped within a public method. The native method is best declared private and its input sanitised by the wrapping method.

https://www.oracle.com/java/technologies/javase/seccodeguide.html#ref-5

"
Guideline 5-3 / INPUT-3: Define wrappers around native methods
Java code is subject to runtime checks for type, array bounds, and library usage. Native code, on the other hand, is generally not. While pure Java code is effectively immune to traditional buffer overflow attacks, native methods are not. To offer some of these protections during the invocation of native code, do not declare a native method public. Instead, declare it private and expose the functionality through a public Java-based wrapper method. A wrapper can safely perform any necessary input validation prior to the invocation of the native method."

**Code Sample:** 

Example code to be flagged

```java
  // public native method
    public native void nativeOperation(byte[] data, int offset,
                                        int len);
```

Example code how to wrap a native method.

```java
public final class NativeMethodWrapper {

    // private native method
    private native void nativeOperation(byte[] data, int offset,
                                        int len);

    // wrapper method performs checks
    public void doOperation(byte[] data, int offset, int len) {
        // copy mutable input
        data = data.clone();

        // validate input
        // Note offset+len would be subject to integer overflow.
        // For instance if offset = 1 and len = Integer.MAX_VALUE,
        //   then offset+len == Integer.MIN_VALUE which is lower
        //   than data.length.
        // Further,
        //   loops of the form
        //       for (int i=offset; i<offset+len; ++i) { ... }
        //   would not throw an exception or cause native code to
        //   crash.

        if (offset < 0 || len < 0 || offset > data.length - len) {
              throw new IllegalArgumentException();
        }

        nativeOperation(data, offset, len);
    }
}
```

**Possible Properties:**
None.

*   Should this rule be customizable via properties?
No. See above.
